packages/ckeditor5-table/src/converters/table-cell-refresh-post-fixer.js

```diff
@@ -25,79 +25,89 @@ export default function injectTableCellRefreshPostFixer( model ) {
 function tableCellRefreshPostFixer( model ) {
 	const differ = model.document.differ;
 
-	// Stores cells to be refreshed so the table cell will be refreshed once for multiple changes.
-	const cellsToRefresh = new Set();
-
-	// Counting the paragraph inserts to verify if it increased to more than one paragraph in the current differ.
-	let insertCount = 0;
+	const changesForCells = new Map();
+	const changes = [ ...differ.getChanges() ];
 
-	for ( const change of differ.getChanges() ) {
-		const parent = change.type == 'insert' || change.type == 'remove' ? change.position.parent : change.range.start.parent;
+	// Updated refresh algorithm.
+	// 1. Gather all changes inside table cell.
+	changes.forEach( change => {
+		const parent = change.type == 'attribute' ? change.range.start.parent : change.position.parent;
 
 		if ( !parent.is( 'element', 'tableCell' ) ) {
-			continue;
+			return;
 		}
 
-		if ( change.type == 'insert' ) {
-			insertCount++;
+		if ( !changesForCells.has( parent ) ) {
+			changesForCells.set( parent, [] );
 		}
 
-		if ( checkRefresh( parent, change.type, insertCount ) ) {
-			cellsToRefresh.add( parent );
+		changesForCells.get( parent ).push( change );
+	} );
+
+	// Stores cells to be refreshed, so the table cell will be refreshed once for multiple changes.
+	const cellsToRefresh = new Set();
+
+	// 2. For each table cell:
+	for ( const [ tableCell, changes ] of changesForCells.entries() ) {
+		// 2a. Count inserts/removes as diff and marks any attribute change.
+		const { childDiff, attribute } = getChangesSummary( changes );
+
+		// 2b. If we detect that number of children has changed...
+		if ( childDiff !== 0 ) {
+			const prevChildren = tableCell.childCount - childDiff;
+			const currentChildren = tableCell.childCount;
+
+			// Might need refresh if previous children was different from 1. Eg.: it was 2 before, now is 1.
+			if ( currentChildren === 1 && prevChildren !== 1 ) {
+				cellsToRefresh.add( tableCell );
+			}
+
+			// Might need refresh if previous children was 1. Eg.: it was 1 before, now is 5.
+			if ( currentChildren !== 1 && prevChildren === 1 ) {
+				cellsToRefresh.add( tableCell );
+			}
```

Comment:
```suggestion
			// Might need refresh if previous children was different from 1. Eg.: it was 2 before, now is 1 (or the opposite).
			if ( currentChildren === 1 || prevChildren === 1 ) {
				cellsToRefresh.add( tableCell );
			}
```

Reply:
I think that this check is too simplistic - a change that adds and removes children would have `currentChildren === 1` and `prevChildren === 1`.

I think that the tests must be updated to check for this case and then let's see.

Reply:
but this scenario is excluded above by 

```
if ( childDiff !== 0 ) {
```

Reply:
I'm talking about:  
Before: `<p>`  
Changes:  `[ add p, remove p ]`  
After: `<p>`  
  
But let's test it to be sure.

Reply:
OMG, I'm not smart with that... there will be no change in the differ itself.